cm/cm_hdd_cdd/hddcdd.py

```diff
@@ -0,0 +1,214 @@
+
+
+@lru_cache(maxsize=256)
+def get_datadir(
+    datarepository: Path,
+    sim_type: str = "historical",
+    dd_type: str = "hdd",
+    Tb=18.0,
+    aggr_window="monthly",
+    method="average",
+) -> Path:
+    """
+    >>> get_datadir("cm/cm_hdd_cdd/testdata").as_posix()
+    "cm/cm_hdd_cdd/testdata/historical/hdd/18.0/monthly/average/"
+    """
+    return (
+        Path(datarepository) / sim_type / dd_type / f"{Tb:.1f}" / aggr_window / method
+    )
+
+
+@lru_cache()
+def extract_by_dir(
+    gdir: Path,
+    lon: float,
+    lat: float,
+    __datasets: Dict[str, rasterio.DatasetReader] = {},
+):
+    res = []
+    idx = []
+    for gfi in gdir.iterdir():
+        if gfi.name.endswith(".tif"):
+            idx.append(gfi.name[:-4])
+            try:
+                gx = __datasets[gfi.as_posix()]
+            except KeyError:
+                gx = rasterio.open(gfi)
+                __datasets[gfi.as_posix()] = gx
+            yp, xp = gx.index(lon, lat)
+            val = gx.read()[0][yp, xp]
+            logging.info(
+                f"{gfi.name}@{lon:.{DECIMALS}f},{lat:.{DECIMALS}f}: {yp}, {xp} => {val}"
+            )
+            res.append(val)
+    sr = pd.Series(np.array(res), index=idx, name=f"yp={yp},xp={xp}")
+    sr.sort_index(inplace=True)
+    return sr
+
+
+def dd_my_group(sr: pd.Series) -> pd.Series:
+    isplit = sr.index.str.split("_")
+    years = [yr for yr, _ in isplit]
+    mnths = [mnth for _, mnth in isplit]
+    yrgrp = sr.groupby(years)
+    mngrp = sr.groupby(mnths)
+    return mngrp, yrgrp
+
+
+def hdd_cdd_stats(
+    geojson: Dict,
+    refyear: int = 2050,
+    rcp: str = "4.5",
+    t_base_h: float = 18.0,
+    t_base_c: float = 22.0,
+):
+    """The `hdd_cdd_stats` returns a set of graphs and KPI extracted by the HDD and CDD
+    layers computed starting from the EURO CORDEX ensamble simulations.
+
+    Graphs​
+    ======
+    This funtion return the following graphs:
+    * the total annual HDD & CDD for the baseline and the future year
+      with confidence interval;
+    * the total monthly HDD & CDD for the baseline and the future year
+      with confidence interval;
+
+    Key Performance Indicator​s (KPIs)
+    =================================
+    The main KPIs provide by the function are:
+    * percentage variation of HDD respect to the current scenario​;
+    * percentage variation of CDD respect to the current scenario​;
+    * reduction of the heating season length​
+    * extension of the cooling season length
+    * number of tropical days
+    """
+    # select the right layer and return the results
+    # start = time()
+    msg = (
+        f"    » ref. year: {refyear},"
+        f" crp: {rcp}, Tbh: {t_base_h:.1f}, Tbc: {t_base_c:.1f}"
+    )
+    logging.info(msg)
+    # Compute the centroid of the geometry selected by the user
+    print(f"CENTROID: {compute_centroid(geojson)}")
+    lon, lat = compute_centroid(geojson)
```

Comment:
The issue here is that I receive a geojson as a Dict, therefore I cannot create a cache because dict in an unhashable type. So I should convert the geojson dict back to string and then I can create a cache, but I think that doing so I'm going to loose most of the time that I might save with the cache. To compute the centroid, I'm just:
* creating a numpy array instance,
* compute the mean values, 
* round the results and 
* transform the output into a tuple.
I think this function should not be that critical. However if you think it is, then we should create a good benchmark function that use real geojson dictionaries from our NUTS{0,1,2,3} and LAU2 to see which is the real gain in adding a cache at this point.

Reply:
aah, now I see what you mean, here I'm calling the function two times, ok, I've removed the call in the print statement, now is called just once.